packages/checkout/widgets-lib/src/widgets/sale/hooks/useCurrency.ts

```diff
@@ -0,0 +1,58 @@
+import { Environment } from '@imtbl/config';
+import { useState, useCallback } from 'react';
+
+type CurrencyResponse = {
+  decimals: number;
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  erc20_address: string;
+  name: string;
+};
+
+const PRIMARY_SALES_API_BASE_URL = {
+  [Environment.SANDBOX]: 'https://api.sandbox.immutable.com/v1/primary-sales',
+  [Environment.PRODUCTION]: 'https://api.immutable.com/v1/primary-sales',
+};
+
+export const useCurrency = ({
+  env,
+  environmentId,
+}: {
+  env: string;
+  environmentId: string;
+}) => {
+  const [currencyError, setCurrencyError] = useState<Error | undefined>(
+    undefined,
+  );
+  const [currencyResponse, setCurrencyResponse] = useState<
+  CurrencyResponse[] | undefined
+  >(undefined);
+
+  const fetchCurrency = useCallback(async () => {
+    setCurrencyError(undefined);
+
+    try {
+      const baseUrl = `${PRIMARY_SALES_API_BASE_URL[env]}/${environmentId}/currency`;
+      const response = await fetch(baseUrl, {
+        method: 'GET',
+        headers: {
+          // eslint-disable-next-line @typescript-eslint/naming-convention
+          'Content-Type': 'application/json',
+        },
+      });
+
+      const data = await response.json();
+      if (!data) {
+        throw new Error('Failed to fetch currency');
+      }
+      setCurrencyResponse(data.currencies);
+
+      return data.currencies;
+    } catch (error) {
+      setCurrencyError(error as Error);
+    }
+
+    return undefined;
```

Comment:
I'm a bit confused with the pattern we're using here - why is this in a `useCallback` and in a hook? And then why are we using `currencyResponse/currencyError ` - I don't see that being consumed anywhere in the code.

The two places that use it are calling `fetchCurrency` and then processing the response and handling errors - if we're going to use a hook the hook should handle all this.

If I understand correctly the currency value only depends on the `env` and `environmentId`, and we only want to call it once per load. What do you think about calling the currency API once on app load inside the `SaleContextProvider` (do any data processing + error handling), and then exposing that value in the provider? The components that consume it can add that as a dependancy.

Reply:
great points, re: useCallback in a hook - the main reason to extract the currency call into a custom hook and using `useCallback` is for code organization and reusability. 

But having `currencyResponse/currencyError` can help with a component that instantiates the hook to reuse the value instead of refetching it. 

I might have overengineered it here, as the currency endpoint is currently being called twice - one for `/sign` and one for smart checkout. Once per load sounds reasonable.

@jhesgodi what are your thoughts on this? having a hook or no hook?

Reply:
Thanks guys, overall having a hook should the right thing, However, I must agree that the behaviour is not the ideal given that is exposing the fetch to the consumer.

From my POV this could be a better way to setup this hook:
- ** Hook goal:** to fetch and store available currencies from API, and expose selected one
- ** Hook name:** base on the goal the name should be `useSelectCurrency` and the returned value `{ tokenAddress }`
- ** Hook behaviour:**  fetch all currencies, and returns the selected one by token name
```
const { currency } = useSelectCurrency({ name: 'USDC' });
const { decimals, erc20Address: fromTokenAddress } = currency;

// Notes:
// Loading should be deferred until currency was set.
// If currency was not found by name, the it should return the first one found (in our current state, this will be trivial as USDC will always be first but also be available to select, however adding the select by name ensure we switch to default a different currency in the future without refactors)

```

Reply:
**Regarding error handling**
I must also agree, this hook should not need to expose the errors to the consumer.
Instead, handle errors internally to default to an available currency.

If no currencies were found (backend fails or returns empty array), the upper View component should handle this by redirecting to SaleErrorView with type: `SaleErrorTypes.INVALID_PARAMETERS`. As this error view will block the flow preventing user from progressing with a misconfigured params.